File: java-context-os/src/main/java/com/owencli/contextos/pipeline/middleware/ReflectMiddleware.java
package com.owencli.contextos.pipeline.middleware;

import com.owencli.contextos.pipeline.PipelineContext;
import com.owencli.contextos.pipeline.PipelineMiddleware;
import com.owencli.contextos.reflection.ReflectionEngine;
import java.util.concurrent.CompletableFuture;

public class ReflectMiddleware implements PipelineMiddleware {
    private final ReflectionEngine r;
    public ReflectMiddleware(ReflectionEngine re) { this.r = re; }
    public String name() { return "reflect"; }
    public int order() { return 800; }
    public CompletableFuture<Void> execute(PipelineContext ctx) {
        return r.reflect(ctx.taskSpec(), ctx.llmResponse(), ctx.metrics().isSuccess(),
            ctx.metrics().isSuccess() ? null : "failure", ctx.userId()).thenAccept(x -> {});
    }
}

File: context_os/pipeline/middlewares/reflect_middleware.py
"""ReflectMiddleware — 后任务自省。"""
from context_os.pipeline.context import PipelineContext
from context_os.pipeline.middleware import PipelineMiddleware
from context_os.memory.experience import ExperienceMemory


class ReflectMiddleware(PipelineMiddleware):
    def __init__(self, experience_memory: ExperienceMemory):
        self._exp = experience_memory

    def name(self):
        return "reflect"

    def order(self):
        return 800

    async def execute(self, ctx: PipelineContext) -> None:
        if not ctx.metrics:
            return
        await self._exp.record_reflection(
            task_type=ctx.task_spec.intent.value if ctx.task_spec else "unknown",
            root_cause="metrics indicated failure" if not ctx.metrics.success else "",
            lesson=ctx.metrics.model_dump_json() if ctx.metrics.success else "",
            success=ctx.metrics.success,
            metadata={"session": ctx.session_id},
        )
